scale_build/packages/git.py

```diff
@@ -45,25 +45,70 @@ def git_log_file(self):
     def git_log_file_path(self):
         return os.path.join(GIT_LOG_DIR, f'{self.name}.log')
 
-    def checkout(self, branch_override=None):
+    def checkout(self, branch_override=None, retries=3):
         origin_url = self.retrieve_current_remote_origin_and_sha()['url']
         branch = branch_override or self.branch
-        if branch == self.existing_branch and self.origin == origin_url:
-            logger.debug('Updating git repo [%s (using branch %s)] (%s)', self.name, branch, self.git_log_file_path)
-            with LoggingContext(self.git_log_file, 'w'):
-                run(['git', '-C', self.source_path, 'fetch', 'origin'])
-                run(['git', '-C', self.source_path, 'checkout', branch])
-                run(['git', '-C', self.source_path, 'reset', '--hard', f'origin/{branch}'])
+        update = (branch == self.existing_branch) and self.origin == origin_url
+        if update:
+            cmds = (
+                ['git', '-C', self.source_path, 'fetch', 'origin'],
+                ['git', '-C', self.source_path, 'checkout', branch],
+                ['git', '-C', self.source_path, 'reset', '--hard', f'origin/{branch}'],
+            )
         else:
-            logger.debug('Checking out git repo [%s (using branch %s)] (%s)', self.name, branch, self.git_log_file_path)
-            if os.path.exists(self.source_path):
-                shutil.rmtree(self.source_path)
+            cmds = (
+                ['git', 'clone', '--recurse', self.origin, self.source_path],
+                ['git', '-C', self.source_path, 'checkout', branch],
+            )
+
+        # We're doing retries here because at the time of writing this the iX network
+        # is having issues with an external hop through the routing of the interwebz
+        # getting to github.com. They've found a particular hop is dropping significant
+        # amounts of packets (~75%+). This is happening network wide so we've got the
+        # retries.
+        # NOTE: when the issue is fixed, we could remove this retry logic
+        retries = 3 if retries <= 0 or retries > 10 else retries
```

Comment:
Do we need this ? And if we do, do you think it's better to raise an exception that less then 0 or bigger then 10 is not allowed rather then implicitly defaulting to 3 ?